UNIXFS.md

```diff
@@ -46,10 +46,15 @@ message Data {
 
 	optional uint64 hashType = 5;
 	optional uint64 fanout = 6;
+
+	repeated Metadata metadata = 7;
```

Comment:
I chose repeated metadata structs like this as it allows us to use one byte to represent every missing piece of metadata. We could alternatively use:

1. A map of integers to metadata.
2. A bitmap to indicate which files have metadata.

However, both feel like overkill. I'm guessing directories will either _have_ metadata or won't.

Ideally, we'd be able to just embed this metadata in the links section of the parent object but that is not to be...

Reply:
Just to be clarify, does the order the metadata appears here maps directly to the ordering of the `Links` in the dag-pb node? i.e. metadata[3] applies to Links[3].

...and this only applies when `Type` is `Directory` right? Can this be documented please?

Also, if we have one or more files in a directory with metadata then _every_ file has to have metadata right so the order doesn't get mixed up?

Separately, did we consider allowing a single metadata property on the root node for each _file_ so that we can still know metadata when the file is accessed directly? If the reason is in order to not change a file's hash because of meta then maybe mention the reason behind this design decision (perhaps this is not something you typically put in a spec, but, where else is appropriate)?

How does this work for HAMT shards? `Links` in a sharded dag-pb node don't map directly to files and secondly listing perms for every file in a sharded directory would make a directory node unfeasibily large.

It's very likely metadata will be used for HAMT shards so any solution we come up with here should work with them. Can we please document the solution?

Reply:
> Also, if we have one or more files in a directory with metadata then every file has to have metadata right so the order doesn't get mixed up?

Yes. I've expanded the spec a bit.

> Separately, did we consider allowing a single metadata property on the root node for each file so that we can still know metadata when the file is accessed directly? If the reason is in order to not change a file's hash because of meta then maybe mention the reason behind this design decision (perhaps this is not something you typically put in a spec, but, where else is appropriate)?

It was considered and dropped for the reason you gave. I'll document it.

> How does this work for HAMT shards? Links in a sharded dag-pb node don't map directly to files and secondly listing perms for every file in a sharded directory would make a directory node unfeasibily large.
> It's very likely metadata will be used for HAMT shards so any solution we come up with here should work with them. Can we please document the solution?

Should be fixed.

Reply:
please no int maps

Reply:
@rvagg could you expand on that? ( I am clearly missing prior discussion of a downside... )

Reply:
This probably isn't the place to litigate this, it always ends in too much text being spilled, but int maps are something we're trying to avoid for IPLD. They're just not straightforward across codecs and programming languages. https://github.com/ipld/specs/blob/master/data-model-layer/data-model.md#motivation